File: trunk/N6LYT/ghpsdr/agc.c
/** 
* @file agc.c
* @brief Functions related to the operation of the AGC. 
* @author John Melton, G0ORX/N6LYT  
* @version 0.1 
* @date 2009-04-11
*/


/* Copyright (C) 
* 2009 - John Melton, G0ORX/N6LYT, Doxygen Comments Dave Larsen, KV0S
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
* 
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with this program; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
* 
*/

// agc.c

#include <gtk/gtk.h>
#include <gdk/gdkkeysyms.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <getopt.h>

#include "agc.h"
#include "bandstack.h"
#include "command.h"
#include "main.h"
#include "property.h"

GtkWidget* agcFixed;

int agc=agcLONG;

GtkWidget* buttonSLOW;
GtkWidget* buttonMEDIUM;
GtkWidget* buttonFAST;
GtkWidget* buttonLONG;

GtkWidget* currentAgcButton;

void setAgc(int agc);

/* --------------------------------------------------------------------------*/
/** 
* @brief Select the agc - sends a setAgc. 
* 
* @param  widget -- Parent widget
*/
void selectAgc(GtkWidget* widget) {
    GtkWidget* label;
    char temp[80];

    if(currentAgcButton) {
        label=gtk_bin_get_child((GtkBin*)currentAgcButton);
        gtk_widget_modify_fg(label, GTK_STATE_NORMAL, &white);
        gtk_widget_modify_fg(label, GTK_STATE_PRELIGHT, &black);
    }
    label=gtk_bin_get_child((GtkBin*)widget);
    gtk_widget_modify_fg(label, GTK_STATE_NORMAL, &buttonSelected);
    gtk_widget_modify_fg(label, GTK_STATE_PRELIGHT, &buttonSelected);
    currentAgcButton=widget;

    if(widget==buttonSLOW) {
        agc=agcSLOW;
    } else if(widget==buttonMEDIUM) {
        agc=agcMEDIUM;
    } else if(widget==buttonFAST) {
        agc=agcFAST;
    } else if(widget==buttonLONG) {
        agc=agcLONG;
    }


    // set RX agc
    sprintf(temp,"setRXAGC %d",agc);
    writeCommand(temp);

}

/* --------------------------------------------------------------------------*/
/** 
* @brief Set the Automatic Gain Control (AGC) 
* 
* @param agc
*/
void setAgc(int agc) {
    GtkWidget* widget;
    switch(agc) {
        case agcSLOW:
            widget=buttonSLOW;
            break;
        case agcMEDIUM:
            widget=buttonMEDIUM;
            break;
        case agcFAST:
            widget=buttonFAST;
            break;
        case agcLONG:
            widget=buttonLONG;
            break;
    }
    selectAgc(widget);
}

/* --------------------------------------------------------------------------*/
/** 
* @brief Callback when a AGC button is pressed.
* 
* @param widget
* @param data
*/
void agcCallback(GtkWidget* widget,gpointer data) {
    selectAgc(widget);
}

/* --------------------------------------------------------------------------*/
/** 
* @brief Build the GUI
* 
* @return 
*/
GtkWidget* buildAgcUI() {
    GtkWidget* label;

    agcFixed=gtk_fixed_new();
    gtk_widget_modify_bg(agcFixed,GTK_STATE_NORMAL,&background);

    // agc buttons
    buttonSLOW = gtk_button_new_with_label ("SLOW");
    gtk_widget_modify_bg(buttonSLOW, GTK_STATE_NORMAL, &buttonBackground);
    label=gtk_bin_get_child((GtkBin*)buttonSLOW);
    gtk_widget_modify_fg(label, GTK_STATE_NORMAL, &white);
    gtk_widget_set_size_request(GTK_WIDGET(buttonSLOW),50,25);
    g_signal_connect(G_OBJECT(buttonSLOW),"clicked",G_CALLBACK(agcCallback),NULL);
    gtk_widget_show(buttonSLOW);
    gtk_fixed_put((GtkFixed*)agcFixed,buttonSLOW,0,0);

    buttonMEDIUM = gtk_button_new_with_label ("MEDIUM");
    gtk_widget_modify_bg(buttonMEDIUM, GTK_STATE_NORMAL, &buttonBackground);
    label=gtk_bin_get_child((GtkBin*)buttonMEDIUM);
    gtk_widget_modify_fg(label, GTK_STATE_NORMAL, &white);
    gtk_widget_set_size_request(GTK_WIDGET(buttonMEDIUM),50,25);
    g_signal_connect(G_OBJECT(buttonMEDIUM),"clicked",G_CALLBACK(agcCallback),NULL);
    gtk_widget_show(buttonMEDIUM);
    gtk_fixed_put((GtkFixed*)agcFixed,buttonMEDIUM,50,0);

    buttonFAST = gtk_button_new_with_label ("FAST");
    gtk_widget_modify_bg(buttonFAST, GTK_STATE_NORMAL, &buttonBackground);
    label=gtk_bin_get_child((GtkBin*)buttonFAST);
    gtk_widget_modify_fg(label, GTK_STATE_NORMAL, &white);
    gtk_widget_set_size_request(GTK_WIDGET(buttonFAST),50,25);
    g_signal_connect(G_OBJECT(buttonFAST),"clicked",G_CALLBACK(agcCallback),NULL);
    gtk_widget_show(buttonFAST);
    gtk_fixed_put((GtkFixed*)agcFixed,buttonFAST,0,25);

    buttonLONG = gtk_button_new_with_label ("LONG");
    gtk_widget_modify_bg(buttonLONG, GTK_STATE_NORMAL, &buttonBackground);
    label=gtk_bin_get_child((GtkBin*)buttonLONG);
    gtk_widget_modify_fg(label, GTK_STATE_NORMAL, &white);
    gtk_widget_set_size_request(GTK_WIDGET(buttonLONG),50,25);
    g_signal_connect(G_OBJECT(buttonLONG),"clicked",G_CALLBACK(agcCallback),NULL);
    gtk_widget_show(buttonLONG);
    gtk_fixed_put((GtkFixed*)agcFixed,buttonLONG,50,25);

    gtk_widget_set_size_request(GTK_WIDGET(agcFixed),100,50);
    gtk_widget_show(agcFixed);

    setAgc(agc);

    return agcFixed;
  
}

/* --------------------------------------------------------------------------*/
/** 
* @brief Save the Automatic Gain Control (AGC) state
*/
void agcSaveState() {
    char string[128];
    sprintf(string,"%d",agc);
    setProperty("agc",string);
}

/* --------------------------------------------------------------------------*/
/** 
* @brief Restore the Automatic Gain Control (AGC) state
*/
void agcRestoreState() {
    char* value;
    value=getProperty("agc");
    if(value) agc=atoi(value);
}

File: trunk/CyclopsII/C# Code/Cyclops/SharpDSP/SharpDSP/Agc.cs
/*
This file is part of a program that implements a Software-Defined Radio.

Copyright (C) 2007, 2008 Philip A Covington

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

The author can be reached by email at

p.covington@gmail.com

*/

using System;

namespace SharpDSP2._1
{
	/// <summary>
	/// Description of agc_new.
	/// </summary>
	public class AGC : IDSPObject
	{
		#region Private members

        private DSPBuffer d = null;
        private DSPState s = null;

        private float setpoint = 1.0f;
        private float error = 0f;
        private float previous_error = 0f;
        private float derivative = 0f;
        private float integral = 0f;
        private float proportional = 0f;
        private float gain = 1.0f;

        private float Kpa = 5.0f; // attack
        private float Kpd = 1500.0f; // decay

        private float mingain = 1.0f;

        #endregion
                          	
        #region Constructor
        
		public AGC(ref DSPBuffer dsp_buffer_obj)
		{
            this.d = dsp_buffer_obj;
            this.s = d.State;                        
		}
		
		#endregion
		
		#region Public Methods
		
        public void Process()
        {
            switch (agcmode)
            {
                case AGCType_e.agcLong:
                    Kpa = 0.5f;
                    Kpd = 675.0f;
                    break;
                case AGCType_e.agcSlow:
                    Kpa = 1.0f;
                    Kpd = 1250.0f;
                    break;
                case AGCType_e.agcMedium:
                    Kpa = 2.0f;
                    Kpd = 2500.0f;
                    break;
                case AGCType_e.agcFast:
                    Kpa = 5.0f;
                    Kpd = 5000.0f;
                    break;
                case AGCType_e.agcOff:
                    d.Scale(fixedgain, fixedgain);
                    return;                    
            }			
        	DoAGC();            
        }
        
        #endregion
        
        #region Private Methods

        private void DoAGC()
        {
            int blocksize = s.DSPBlockSize;

            float envelope = d.GetBlockAvgMagnitudeWithGain(gain, blocksize);

            previous_error = error;
            
            // compute error
            error = setpoint - envelope;

            // deadband
            if ((float)Math.Abs(error) < 0.1f)
            {
                integral = 0f;
                d.Scale(gain, gain);
                gain -= gain_leakage;
                return;
            }

            // computer proportional term
            if (error < 0f)
            {
                proportional = error * Kpd;  // gain decreasing
            }
            else
            {
                proportional = error * Kpa; // gain increasing
            }
            
            //compute integral term
            integral += error * Ki;

            if (integral > integral_limit)
                integral = integral_limit;

            // compute derivative term
            derivative = (error - previous_error) * Kd;

            float Po = proportional + integral + derivative;

            gain += Po;
            if (gain >= maxgain)
                gain = maxgain;
            if (gain < mingain)
                gain = mingain;

            // apply sgc
            d.Scale(gain, gain);

            // gain leakage
            gain -= gain_leakage;
        }                       
                        
        #endregion

        #region Properties

        private float Ki = 10f;
        private float Ki_log = 10f;
        public float AGCIntegralGainDB
        {
            get { return Ki_log; }
            set 
            { 
                Ki_log = value;
                Ki = (float)Math.Pow(10.0, Ki_log / 20.0);
            }
        }

        private float Kd = 50f;
        private float Kd_log = 20f;
        public float AGCDifferentialGainDB
        {
            get { return Kd_log; }
            set 
            { 
                Kd_log = value;
                Kd = (float)Math.Pow(10.0, Kd_log / 20.0);
            }
        }

        private float maxgain = 31622f;
        private float maxloggain = 90f;
        public float AGCMaximumGainDB
        {
            get { return maxloggain; }
            set 
            { 
                maxloggain = value;
                maxgain = (float)Math.Pow(10.0, maxloggain / 20.0);
            }
        }

        private float fixedgain = 316f;
        private float fixedloggain = 50f;
        public float AGCFixedGainDB
        {
            get { return fixedloggain; }
            set
            {
                fixedloggain = value;
                fixedgain = (float)Math.Pow(10.0, fixedloggain / 20.0);
            }
        }

        private float integral_limit = 100f;
        private float integral_log_limit = 20f;
        public float AGCIntegralLimitDB
        {
            get { return integral_log_limit; }
            set
            {
                integral_log_limit = value;
                integral_limit = (float)Math.Pow(10.0, integral_log_limit / 20.0);
            }
        }

        private float gain_leakage = 4f; 
        private float gain_log_leakage = 12f;
        public float AGCGainLeakageDB
        {
            get { return gain_log_leakage; }
            set
            {
                gain_log_leakage = value;
                gain_leakage = (float)Math.Pow(10.0, gain_log_leakage / 20.0);
            }
        }

        private AGCType_e agcmode = AGCType_e.agcMedium;
        public AGCType_e AGCMode
        {
            get { return agcmode; }
            set { agcmode = value; }
        }

        #endregion
    }
}
